### PR TITLE
[batch] Don't rmtree if any errors occur while unmounting

### PR DESCRIPTION
**Pull request by jigold**
I think this change is better than #12975

### CODE REVIEW

**Comment by jigold** [comment]
It would be great in a separate PR to check for sure there aren't any zombie fuse mounts before removing files.

File: batch/batch/worker/worker.py [hunk 1]
```diff
@@ -2212,25 +2213,31 @@ async def cleanup(self):
...
                     mount_path = self.cloudfuse_data_path(bucket)
-                    await self.jvm.cloudfuse_mount_manager.unmount(mount_path, user=self.user, bucket=bucket)
-                    config['mounted'] = False
+                    try:
+                        await self.jvm.cloudfuse_mount_manager.unmount(mount_path, user=self.user, bucket=bucket)
+                        config['mounted'] = False
+                    except asyncio.CancelledError:
+                        raise
+                    except Exception:
+                        log.exception(f'error while unmounting cloudfuse for {self.jvm_name} for job {self.id}')
```
Comment (danking) [inline batch/batch/worker/worker.py:2234]: Let's include the same information we have above:
```suggestion
                        log.exception(
                            f'while unmounting fuse blob storage {bucket} from {mount_path} for {self.jvm_name} for job {self.id}'
                        )
```

File: batch/batch/worker/worker.py [hunk 2]
```diff
@@ -2771,6 +2788,10 @@ async def run_job(self, job):
             raise
         except JVMCreationError:
             self.stop_event.set()
+        except IncompleteJVMCleanupError:
+            assert isinstance(job, JVMJob)
+            await self.recreate_jvm(job.jvm)
+            log.exception(f'while running {job}, incomplete cleanup from {job.jvm}')
```
Comment (danking) [inline batch/batch/worker/worker.py:2794]: This results in two exceptional logs for a single exception. Let's either not log here or not log near the raise.

**Review by danking — CHANGES_REQUESTED** [review]

File: batch/batch/worker/worker.py [hunk 3]
```diff
@@ -2231,8 +2231,9 @@ async def cleanup(self):
                     except asyncio.CancelledError:
                         raise
                     except Exception:
-                        log.exception(f'error while unmounting cloudfuse for {self.jvm_name} for job {self.id}')
-                        raise IncompleteJVMCleanupError
+                        raise IncompleteJVMCleanupError(
+                            f'while unmounting fuse blob storage {bucket} from {mount_path} for {self.jvm_name} for job {self.id}'
+                        )
```
Comment (danking) [inline batch/batch/worker/worker.py:2236]: Name the exception and add a `from e` so that we get the fulls tack trace in the log.exception later.

**Review by danking — CHANGES_REQUESTED** [review]
x

File: batch/batch/worker/worker.py [hunk 4]
```diff
@@ -2312,6 +2320,11 @@ class JVMCreationError(Exception):
     pass
 
 
+class IncompleteJVMCleanupError(Exception):
+    def __init__(self, msg):
+        self.msg = msg
```
Comment (danking) [inline batch/batch/worker/worker.py:2325]: This isn't necessary:
```
In [1]: class IncompleteJVMCleanupError(Exception):
   ...:     pass
   ...: 

In [2]: raise IncompleteJVMCleanupError('foo')
---------------------------------------------------------------------------
IncompleteJVMCleanupError                 Traceback (most recent call last)
<ipython-input-2-a8a0c94b0ce4> in <cell line: 1>()
----> 1 raise IncompleteJVMCleanupError('foo')

IncompleteJVMCleanupError: foo
```

**Review by danking — DISMISSED** [review]

**Comment by daniel-goldstein** [comment]
These changes look great

**Review by danking — APPROVED** [review]
LGTM!

File: batch/batch/worker/worker.py [hunk 5]
```diff
@@ -2789,10 +2789,10 @@ async def run_job(self, job):
             raise
         except JVMCreationError:
             self.stop_event.set()
-        except IncompleteJVMCleanupError as e:
+        except IncompleteJVMCleanupError:
             assert isinstance(job, JVMJob)
             await self.recreate_jvm(job.jvm)
-            log.exception(e.msg)
+            log.exception(f'while running {job}, ignoring')
```
[the remark this answers is not in the document]
Comment (danking) [reply batch/batch/worker/worker.py:2796]: Yeah, [`log.exception` does this](https://docs.python.org/3/library/logging.html#logging.Logger.exception).

**Review by danking — APPROVED** [review]

**Pull request merged**